{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "suppressMessages(library(ggplot2))\n",
    "suppressMessages(library(ArchR))\n",
    "suppressMessages(library(patchwork))\n",
    "suppressMessages(library(BSgenome.Hsapiens.UCSC.hg38))\n",
    "suppressMessages(library(dplyr))\n",
    "suppressMessages(library(cowplot))\n",
    "suppressMessages(library(ggrepel))\n",
    "suppressMessages(library(rhdf5))\n",
    "suppressMessages(library(chromVARmotifs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Setting default number of Parallel threads to 64.\n",
      "\n",
      "Setting default genome to Hg38.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "set.seed(42)\n",
    "addArchRThreads(threads = 64)\n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Successfully loaded ArchRProject!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "proj <- loadArchRProject(\"./ArchRProject\", showLogo = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "86017"
      ],
      "text/latex": [
       "86017"
      ],
      "text/markdown": [
       "86017"
      ],
      "text/plain": [
       "[1] 86017"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "length(proj$cellNames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addGroupCoverages-c314c33301f16-Date-2025-02-24_Time-12-26-52.408171.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "10N (1 of 11) : CellGroups N = 2\n",
      "\n",
      "10T (2 of 11) : CellGroups N = 2\n",
      "\n",
      "11T (3 of 11) : CellGroups N = 2\n",
      "\n",
      "12N (4 of 11) : CellGroups N = 2\n",
      "\n",
      "12T (5 of 11) : CellGroups N = 2\n",
      "\n",
      "13T (6 of 11) : CellGroups N = 2\n",
      "\n",
      "18T (7 of 11) : CellGroups N = 2\n",
      "\n",
      "19T (8 of 11) : CellGroups N = 2\n",
      "\n",
      "20T (9 of 11) : CellGroups N = 2\n",
      "\n",
      "21T (10 of 11) : CellGroups N = 2\n",
      "\n",
      "22T (11 of 11) : CellGroups N = 2\n",
      "\n",
      "2025-02-24 12:26:57.540692 : Further Sampled 22 Groups above the Max Fragments!, 0.086 mins elapsed.\n",
      "\n",
      "2025-02-24 12:26:58.094321 : Creating Coverage Files!, 0.095 mins elapsed.\n",
      "\n",
      "2025-02-24 12:26:58.097149 : Batch Execution w/ safelapply!, 0.095 mins elapsed.\n",
      "\n",
      "2025-02-24 12:26:58.150272 : Group 10N._.Rep1 (1 of 22) : Creating Group Coverage File : X10N._.Rep1.insertions.coverage.h5, 0.096 mins elapsed.\n",
      "\n",
      "Number of Cells = 2545\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:27:52.585872 : Group 10N._.Rep2 (2 of 22) : Creating Group Coverage File : X10N._.Rep2.insertions.coverage.h5, 1.003 mins elapsed.\n",
      "\n",
      "Number of Cells = 2567\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:28:44.782039 : Group 10T._.Rep1 (3 of 22) : Creating Group Coverage File : X10T._.Rep1.insertions.coverage.h5, 1.873 mins elapsed.\n",
      "\n",
      "Number of Cells = 2322\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:29:36.050537 : Group 10T._.Rep2 (4 of 22) : Creating Group Coverage File : X10T._.Rep2.insertions.coverage.h5, 2.727 mins elapsed.\n",
      "\n",
      "Number of Cells = 2277\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:30:27.412084 : Group 11T._.Rep1 (5 of 22) : Creating Group Coverage File : X11T._.Rep1.insertions.coverage.h5, 3.583 mins elapsed.\n",
      "\n",
      "Number of Cells = 3338\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:31:12.293638 : Group 11T._.Rep2 (6 of 22) : Creating Group Coverage File : X11T._.Rep2.insertions.coverage.h5, 4.331 mins elapsed.\n",
      "\n",
      "Number of Cells = 3344\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:31:56.825153 : Group 12N._.Rep1 (7 of 22) : Creating Group Coverage File : X12N._.Rep1.insertions.coverage.h5, 5.074 mins elapsed.\n",
      "\n",
      "Number of Cells = 3744\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:32:41.67062 : Group 12N._.Rep2 (8 of 22) : Creating Group Coverage File : X12N._.Rep2.insertions.coverage.h5, 5.821 mins elapsed.\n",
      "\n",
      "Number of Cells = 3734\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:33:28.107184 : Group 12T._.Rep1 (9 of 22) : Creating Group Coverage File : X12T._.Rep1.insertions.coverage.h5, 6.595 mins elapsed.\n",
      "\n",
      "Number of Cells = 3622\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:34:14.241788 : Group 12T._.Rep2 (10 of 22) : Creating Group Coverage File : X12T._.Rep2.insertions.coverage.h5, 7.364 mins elapsed.\n",
      "\n",
      "Number of Cells = 3649\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:34:59.096927 : Group 13T._.Rep1 (11 of 22) : Creating Group Coverage File : X13T._.Rep1.insertions.coverage.h5, 8.111 mins elapsed.\n",
      "\n",
      "Number of Cells = 3101\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:35:49.297117 : Group 13T._.Rep2 (12 of 22) : Creating Group Coverage File : X13T._.Rep2.insertions.coverage.h5, 8.948 mins elapsed.\n",
      "\n",
      "Number of Cells = 3085\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:36:37.041537 : Group 18T._.Rep1 (13 of 22) : Creating Group Coverage File : X18T._.Rep1.insertions.coverage.h5, 9.744 mins elapsed.\n",
      "\n",
      "Number of Cells = 1791\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:37:26.231616 : Group 18T._.Rep2 (14 of 22) : Creating Group Coverage File : X18T._.Rep2.insertions.coverage.h5, 10.564 mins elapsed.\n",
      "\n",
      "Number of Cells = 1771\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:38:14.044234 : Group 19T._.Rep1 (15 of 22) : Creating Group Coverage File : X19T._.Rep1.insertions.coverage.h5, 11.361 mins elapsed.\n",
      "\n",
      "Number of Cells = 3081\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:38:59.652575 : Group 19T._.Rep2 (16 of 22) : Creating Group Coverage File : X19T._.Rep2.insertions.coverage.h5, 12.121 mins elapsed.\n",
      "\n",
      "Number of Cells = 3094\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:39:45.083951 : Group 20T._.Rep1 (17 of 22) : Creating Group Coverage File : X20T._.Rep1.insertions.coverage.h5, 12.878 mins elapsed.\n",
      "\n",
      "Number of Cells = 2573\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:40:32.227841 : Group 20T._.Rep2 (18 of 22) : Creating Group Coverage File : X20T._.Rep2.insertions.coverage.h5, 13.664 mins elapsed.\n",
      "\n",
      "Number of Cells = 2554\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:41:20.156834 : Group 21T._.Rep1 (19 of 22) : Creating Group Coverage File : X21T._.Rep1.insertions.coverage.h5, 14.462 mins elapsed.\n",
      "\n",
      "Number of Cells = 2852\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:42:07.314948 : Group 21T._.Rep2 (20 of 22) : Creating Group Coverage File : X21T._.Rep2.insertions.coverage.h5, 15.248 mins elapsed.\n",
      "\n",
      "Number of Cells = 2846\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:42:55.468274 : Group 22T._.Rep1 (21 of 22) : Creating Group Coverage File : X22T._.Rep1.insertions.coverage.h5, 16.051 mins elapsed.\n",
      "\n",
      "Number of Cells = 2255\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:43:41.424153 : Group 22T._.Rep2 (22 of 22) : Creating Group Coverage File : X22T._.Rep2.insertions.coverage.h5, 16.817 mins elapsed.\n",
      "\n",
      "Number of Cells = 2234\n",
      "\n",
      "Coverage File Exists!\n",
      "\n",
      "Added Coverage Group\n",
      "\n",
      "Added Metadata Group\n",
      "\n",
      "Added ArrowCoverage Class\n",
      "\n",
      "Added Coverage/Info\n",
      "\n",
      "Added Coverage/Info/CellNames\n",
      "\n",
      "2025-02-24 12:44:29.915774 : Adding Kmer Bias to Coverage Files!, 17.625 mins elapsed.\n",
      "\n",
      "Completed Kmer Bias Calculation\n",
      "\n",
      "Adding Kmer Bias (1 of 22)\n",
      "\n",
      "Adding Kmer Bias (2 of 22)\n",
      "\n",
      "Adding Kmer Bias (3 of 22)\n",
      "\n",
      "Adding Kmer Bias (4 of 22)\n",
      "\n",
      "Adding Kmer Bias (5 of 22)\n",
      "\n",
      "Adding Kmer Bias (6 of 22)\n",
      "\n",
      "Adding Kmer Bias (7 of 22)\n",
      "\n",
      "Adding Kmer Bias (8 of 22)\n",
      "\n",
      "Adding Kmer Bias (9 of 22)\n",
      "\n",
      "Adding Kmer Bias (10 of 22)\n",
      "\n",
      "Adding Kmer Bias (11 of 22)\n",
      "\n",
      "Adding Kmer Bias (12 of 22)\n",
      "\n",
      "Adding Kmer Bias (13 of 22)\n",
      "\n",
      "Adding Kmer Bias (14 of 22)\n",
      "\n",
      "Adding Kmer Bias (15 of 22)\n",
      "\n",
      "Adding Kmer Bias (16 of 22)\n",
      "\n",
      "Adding Kmer Bias (17 of 22)\n",
      "\n",
      "Adding Kmer Bias (18 of 22)\n",
      "\n",
      "Adding Kmer Bias (19 of 22)\n",
      "\n",
      "Adding Kmer Bias (20 of 22)\n",
      "\n",
      "Adding Kmer Bias (21 of 22)\n",
      "\n",
      "Adding Kmer Bias (22 of 22)\n",
      "\n",
      "2025-02-24 12:45:53.06807 : Finished Creation of Coverage Files!, 19.011 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addGroupCoverages-c314c33301f16-Date-2025-02-24_Time-12-26-52.408171.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## 计算覆盖度\n",
    "### 覆盖度：计算每个样本中的reads对基因组的覆盖度，表示每个细胞的测序深度。\n",
    "### maxCells和minCells，限制每个样本中用于计算覆盖度的细胞（超过5000个时，随机选择5000个来计算覆盖度）\n",
    "proj <- addGroupCoverages(ArchRProj = proj, groupBy = \"Sample\",\n",
    "                          maxCells = 5000,\n",
    "                          minCells = 5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "addReproduciblePeakSet          package:ArchR          R Documentation\n",
      "\n",
      "_\bA_\bd_\bd _\ba _\bR_\be_\bp_\br_\bo_\bd_\bu_\bc_\bi_\bb_\bl_\be _\bP_\be_\ba_\bk _\bS_\be_\bt _\bt_\bo _\ba_\bn _\bA_\br_\bc_\bh_\bR_\bP_\br_\bo_\bj_\be_\bc_\bt\n",
      "\n",
      "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
      "\n",
      "     This function will get insertions from coverage files, call peaks,\n",
      "     and merge peaks to get a \"Union Reproducible Peak Set\".\n",
      "\n",
      "_\bU_\bs_\ba_\bg_\be:\n",
      "\n",
      "     addReproduciblePeakSet(\n",
      "       ArchRProj = NULL,\n",
      "       groupBy = \"Clusters\",\n",
      "       peakMethod = \"Macs2\",\n",
      "       reproducibility = \"2\",\n",
      "       peaksPerCell = 500,\n",
      "       maxPeaks = 150000,\n",
      "       minCells = 25,\n",
      "       excludeChr = c(\"chrM\", \"chrY\"),\n",
      "       pathToMacs2 = if (tolower(peakMethod) == \"macs2\") findMacs2() else NULL,\n",
      "       genomeSize = NULL,\n",
      "       shift = -75,\n",
      "       extsize = 150,\n",
      "       method = if (tolower(peakMethod) == \"macs2\") \"q\" else \"p\",\n",
      "       cutOff = 0.1,\n",
      "       additionalParams = \"--nomodel --nolambda\",\n",
      "       extendSummits = 250,\n",
      "       promoterRegion = c(2000, 100),\n",
      "       genomeAnnotation = getGenomeAnnotation(ArchRProj),\n",
      "       geneAnnotation = getGeneAnnotation(ArchRProj),\n",
      "       plot = TRUE,\n",
      "       threads = getArchRThreads(),\n",
      "       parallelParam = NULL,\n",
      "       force = FALSE,\n",
      "       verbose = TRUE,\n",
      "       logFile = createLogFile(\"addReproduciblePeakSet\"),\n",
      "       ...\n",
      "     )\n",
      "     \n",
      "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
      "\n",
      "ArchRProj: An ‘ArchRProject’ object.\n",
      "\n",
      " groupBy: The name of the column in ‘cellColData’ to use for grouping\n",
      "          cells together for peak calling.\n",
      "\n",
      "peakMethod: The name of peak calling method to be used. Options include\n",
      "          \"Macs2\" for using macs2 callpeak or \"Tiles\" for using a\n",
      "          TileMatrix.\n",
      "\n",
      "reproducibility: A string that indicates how peak reproducibility\n",
      "          should be handled. This string is dynamic and can be a\n",
      "          function of ‘n’ where ‘n’ is the number of samples being\n",
      "          assessed. For example, ‘reproducibility = \"2\"’ means at least\n",
      "          2 samples must have a peak call at this locus and\n",
      "          ‘reproducibility = \"(n+1)/2\"’ means that the majority of\n",
      "          samples must have a peak call at this locus.\n",
      "\n",
      "peaksPerCell: The upper limit of the number of peaks that can be\n",
      "          identified per cell-grouping in ‘groupBy’. This is useful for\n",
      "          controlling how many peaks can be called from cell groups\n",
      "          with low cell numbers.\n",
      "\n",
      "maxPeaks: A numeric threshold for the maximum peaks to retain per group\n",
      "          from ‘groupBy’ in the union reproducible peak set.\n",
      "\n",
      "minCells: The minimum allowable number of unique cells that was used to\n",
      "          create the coverage files on which peaks are called. This is\n",
      "          important to allow for exclusion of pseudo-bulk replicates\n",
      "          derived from very low cell numbers.\n",
      "\n",
      "excludeChr: A character vector containing the ‘seqnames’ of the\n",
      "          chromosomes that should be excluded from peak calling.\n",
      "\n",
      "pathToMacs2: The full path to the MACS2 executable.\n",
      "\n",
      "genomeSize: The genome size to be used for MACS2 peak calling (see\n",
      "          MACS2 documentation). This is required if genome is not hg19,\n",
      "          hg38, mm9, or mm10.\n",
      "\n",
      "   shift: The number of basepairs to shift each Tn5 insertion. When\n",
      "          combined with ‘extsize’ this allows you to create proper\n",
      "          fragments, centered at the Tn5 insertion site, for use with\n",
      "          MACS2 (see MACS2 documentation).\n",
      "\n",
      " extsize: The number of basepairs to extend the MACS2 fragment after\n",
      "          ‘shift’ has been applied. When combined with ‘extsize’ this\n",
      "          allows you to create proper fragments, centered at the Tn5\n",
      "          insertion site, for use with MACS2 (see MACS2 documentation).\n",
      "\n",
      "  method: The method to use for significance testing in MACS2. Options\n",
      "          are \"p\" for p-value and \"q\" for q-value. When combined with\n",
      "          ‘cutOff’ this gives the method and significance threshold for\n",
      "          peak calling (see MACS2 documentation).\n",
      "\n",
      "  cutOff: The numeric significance cutOff for the testing method\n",
      "          indicated by ‘method’ (see MACS2 documentation).\n",
      "\n",
      "additionalParams: A string of additional parameters to pass to MACS2\n",
      "          (see MACS2 documentation).\n",
      "\n",
      "extendSummits: The number of basepairs to extend peak summits (in both\n",
      "          directions) to obtain final fixed-width peaks. For example,\n",
      "          ‘extendSummits = 250’ will create 501-bp fixed-width peaks\n",
      "          from the 1-bp summits.\n",
      "\n",
      "promoterRegion: A vector of two integers specifying the distance in\n",
      "          basepairs upstream and downstream of a TSS to be included as\n",
      "          a promoter region. Peaks called within one of these regions\n",
      "          will be annotated as a \"promoter\" peak. For example,\n",
      "          ‘promoterRegion = c(2000, 100)’ will annotate any peak within\n",
      "          the region 2000 bp upstream and 100 bp downstream of a TSS as\n",
      "          a \"promoter\" peak.\n",
      "\n",
      "genomeAnnotation: The genomeAnnotation (see ‘createGenomeAnnotation()’)\n",
      "          to be used for generating peak metadata such as nucleotide\n",
      "          information (GC content) or chromosome sizes.\n",
      "\n",
      "geneAnnotation: The geneAnnotation (see ‘createGeneAnnotation()’) to be\n",
      "          used for labeling peaks as \"promoter\", \"exonic\", etc.\n",
      "\n",
      "    plot: A boolean describing whether to plot peak annotation results.\n",
      "\n",
      " threads: The number of threads to be used for parallel computing.\n",
      "\n",
      "parallelParam: A list of parameters to be passed for\n",
      "          biocparallel/batchtools parallel computing.\n",
      "\n",
      "   force: A boolean value indicating whether to force the reproducible\n",
      "          peak set to be overwritten if it already exist in the given\n",
      "          ‘ArchRProject’ peakSet.\n",
      "\n",
      " verbose: A boolean value that determines whether standard output\n",
      "          includes verbose sections.\n",
      "\n",
      " logFile: The path to a file to be used for logging ArchR output.\n",
      "\n",
      "     ...: Additional parameters to be pass to ‘addGroupCoverages()’ to\n",
      "          get sample-guided pseudobulk cell-groupings. Only used for\n",
      "          TileMatrix-based peak calling (not for MACS2). See\n",
      "          ‘addGroupCoverages()’ for more info.\n",
      "\n",
      "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
      "\n",
      "     # Get Test ArchR Project\n",
      "     proj <- getTestProject()\n",
      "     \n",
      "     # Add Peak Matrix Tiles\n",
      "     proj <- addReproduciblePeakSet(proj, peakMethod = \"tiles\")\n",
      "     \n",
      "     # Add Peak Matrix Macs2 (Preferred)\n",
      "     proj <- addReproduciblePeakSet(proj, peakMethod = \"macs2\")\n",
      "     "
     ]
    }
   ],
   "source": [
    "?addReproduciblePeakSet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Searching For MACS2..\n",
      "\n",
      "Found with $PATH at /home/hanxue/miniconda3/bin/macs2\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addReproduciblePeakSet-c314c659ab309-Date-2025-02-24_Time-13-16-08.279945.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "Calling Peaks with Macs2\n",
      "\n",
      "2025-02-24 13:16:08.421258 : Peak Calling Parameters!, 0.002 mins elapsed.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Group nCells nCellsUsed nReplicates nMin nMax maxPeaks\n",
      "10N   10N  10296      10000           2 5000 5000   150000\n",
      "10T   10T  10218      10000           2 5000 5000   150000\n",
      "11T   11T   8644       7178           2 5000 5000   150000\n",
      "12N   12N   7346       6655           2 5000 5000   150000\n",
      "12T   12T   7089       6516           2 5000 5000   150000\n",
      "13T   13T   9695       7500           2 5000 5000   150000\n",
      "18T   18T   6068       4935           2 3393 3397   150000\n",
      "19T   19T   6367       6103           2 5000 5000   150000\n",
      "20T   20T   7036       6486           2 5000 5000   150000\n",
      "21T   21T   7774       6832           2 5000 5000   150000\n",
      "22T   22T   5484       4637           2 3256 3272   150000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-24 13:16:08.43061 : Batching Peak Calls!, 0.003 mins elapsed.\n",
      "\n",
      "2025-02-24 13:16:08.438231 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "2025-02-24 13:19:58.590448 : Identifying Reproducible Peaks!, 3.838 mins elapsed.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n",
      "R_zmq_msg_send errno: 4 strerror: Interrupted system call\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-24 13:20:08.031703 : Creating Union Peak Set!, 3.996 mins elapsed.\n",
      "\n",
      "Converged after 9 iterations!\n",
      "\n",
      "Plotting Ggplot!\n",
      "\n",
      "2025-02-24 13:20:19.905683 : Finished Creating Union Peak Set (345310)!, 4.194 mins elapsed.\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addPeakMatrix-c314c47033639-Date-2025-02-24_Time-13-20-19.909908.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-24 13:20:20.01029 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addPeakMatrix-c314c47033639-Date-2025-02-24_Time-13-20-19.909908.log\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## peak calling\n",
    "## 使用macs2进行peak calling\n",
    "pathToMacs2 <- findMacs2()\n",
    "\n",
    "## 添加一个可重复的峰值集\n",
    "## 对每个样本进行峰值调用，然后将峰值集合并，过滤掉不可重复的峰值（cutoff）\n",
    "## cutoff，峰值的p值阈值，用于过滤峰值\n",
    "proj <- addReproduciblePeakSet(\n",
    "    ArchRProj = proj, \n",
    "    groupBy = \"Sample\", \n",
    "    cutOff = 0.05,\n",
    "    minCells = 25,\n",
    "    promoterRegion = c(2000, 100),\n",
    "    pathToMacs2 = pathToMacs2\n",
    ")\n",
    "\n",
    "proj <- addPeakMatrix(proj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-getMatrixFromProject-c314c63da3125-Date-2025-02-24_Time-13-28-45.656913.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2025-02-24 13:29:57.77394 : Organizing colData, 1.202 mins elapsed.\n",
      "\n",
      "2025-02-24 13:29:58.114991 : Organizing rowData, 1.208 mins elapsed.\n",
      "\n",
      "2025-02-24 13:29:58.131417 : Organizing rowRanges, 1.208 mins elapsed.\n",
      "\n",
      "2025-02-24 13:29:58.153352 : Organizing Assays (1 of 1), 1.208 mins elapsed.\n",
      "\n",
      "2025-02-24 13:30:40.266477 : Constructing SummarizedExperiment, 1.91 mins elapsed.\n",
      "\n",
      "2025-02-24 13:31:50.698181 : Finished Matrix Creation, 3.084 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## save peak matrix\n",
    "## 峰值矩阵peakMatrix中的计数矩阵counts\n",
    "## df_rangers，峰值的坐标信息\n",
    "peakMatrix <- getMatrixFromProject(proj,\n",
    "                                   useMatrix = \"PeakMatrix\")\n",
    "\n",
    "counts <- peakMatrix@assays@data$PeakMatrix\n",
    "df_rangers <- as.data.frame(peakMatrix@rowRanges@ranges)\n",
    "\n",
    "## 格式化counts的行名为染色体_起始位置_终止位置\n",
    "rownames(counts) <- paste(peakMatrix@rowRanges@seqnames,\n",
    "                          df_rangers$start,\n",
    "                          df_rangers$end,\n",
    "                          sep = \"_\") \n",
    "# 保存峰值矩阵\n",
    "saveRDS(counts, file = \"./ArchRProject/PeakMatrix.Rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-getMatrixFromProject-c314c3eb91604-Date-2025-02-24_Time-13-47-17.04584.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-24 13:48:18.62827 : Organizing colData, 1.026 mins elapsed.\n",
      "\n",
      "2025-02-24 13:48:19.0466 : Organizing rowData, 1.033 mins elapsed.\n",
      "\n",
      "2025-02-24 13:48:19.054247 : Organizing rowRanges, 1.033 mins elapsed.\n",
      "\n",
      "2025-02-24 13:48:19.061831 : Organizing Assays (1 of 1), 1.034 mins elapsed.\n",
      "\n",
      "2025-02-24 13:49:08.418415 : Constructing SummarizedExperiment, 1.856 mins elapsed.\n",
      "\n",
      "2025-02-24 13:49:10.354281 : Finished Matrix Creation, 1.888 mins elapsed.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## save gene matrix\n",
    "## 和峰值矩阵类似，基因矩阵GeneScoreMatrix中的计数矩阵counts，保存基因矩阵\n",
    "atac <- getMatrixFromProject(ArchRProj = proj,\n",
    "                             useMatrix = \"GeneScoreMatrix\")\n",
    "\n",
    "counts <- atac@assays@data$GeneScoreMatrix\n",
    "rownames(counts) <- atac@elementMetadata$name\n",
    "\n",
    "saveRDS(counts, file = \"./ArchRProject/GeneScoreMatrix.Rds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ArchR logging to : ArchRLogs/ArchR-addMotifAnnotations-c314c56366810-Date-2025-02-24_Time-14-18-33.111154.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-24 14:18:33.712746 : Getting Motif Set, Species : , 0.002 mins elapsed.\n",
      "\n",
      "2025-02-24 14:18:35.92506 : Finding Motif Positions with motifmatchr!, 0.039 mins elapsed.\n",
      "\n",
      "2025-02-24 14:23:28.990012 : All Motifs Overlap at least 1 peak!, 4.923 mins elapsed.\n",
      "\n",
      "2025-02-24 14:23:29.024291 : Creating Motif Overlap Matrix, 4.924 mins elapsed.\n",
      "\n",
      "2025-02-24 14:23:33.343299 : Finished Getting Motif Info!, 4.996 mins elapsed.\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addMotifAnnotations-c314c56366810-Date-2025-02-24_Time-14-18-33.111154.log\n",
      "\n",
      "Identifying Background Peaks!\n",
      "\n",
      "Using Previous Background Peaks!\n",
      "\n",
      "ArchR logging to : ArchRLogs/ArchR-addDeviationsMatrix-c314c35b6cf0d-Date-2025-02-24_Time-14-24-56.204333.log\n",
      "If there is an issue, please report to github with logFile!\n",
      "\n",
      "2025-02-24 14:25:14.170182 : Batch Execution w/ safelapply!, 0 mins elapsed.\n",
      "\n",
      "###########\n",
      "2025-02-24 15:01:49.770331 : Completed Computing Deviations!, 36.893 mins elapsed.\n",
      "###########\n",
      "\n",
      "ArchR logging successful to : ArchRLogs/ArchR-addDeviationsMatrix-c314c35b6cf0d-Date-2025-02-24_Time-14-24-56.204333.log\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GeneScoreMatrix'</li><li>'PeakMatrix'</li><li>'TileMatrix'</li><li>'homerMatrix'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'GeneScoreMatrix'\n",
       "\\item 'PeakMatrix'\n",
       "\\item 'TileMatrix'\n",
       "\\item 'homerMatrix'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'GeneScoreMatrix'\n",
       "2. 'PeakMatrix'\n",
       "3. 'TileMatrix'\n",
       "4. 'homerMatrix'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"GeneScoreMatrix\" \"PeakMatrix\"      \"TileMatrix\"      \"homerMatrix\"    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## 添加motif注释\n",
    "proj <- addMotifAnnotations(proj, motifSet = \"homer\", annoName = \"homer\")\n",
    "# 添加背景峰值，添加峰值偏差矩阵\n",
    "## 峰值偏差矩阵\n",
    "proj <- addBgdPeaks(proj) %>% addDeviationsMatrix(., peakAnnotation = \"homer\")\n",
    "# 检查当前ArchRProject中存放了哪些矩阵数据\n",
    "getAvailableMatrices(proj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Saving ArchRProject...\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "R version 4.4.1 (2024-06-14)\n",
       "Platform: x86_64-conda-linux-gnu\n",
       "Running under: Ubuntu 22.04.2 LTS\n",
       "\n",
       "Matrix products: default\n",
       "BLAS/LAPACK: /home/hanxue/miniconda3/envs/scseq/lib/libopenblasp-r0.3.21.so;  LAPACK version 3.9.0\n",
       "\n",
       "Random number generation:\n",
       " RNG:     L'Ecuyer-CMRG \n",
       " Normal:  Inversion \n",
       " Sample:  Rejection \n",
       " \n",
       "locale:\n",
       " [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       \n",
       " [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   \n",
       " [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          \n",
       "[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   \n",
       "\n",
       "time zone: Etc/UTC\n",
       "tzcode source: system (glibc)\n",
       "\n",
       "attached base packages:\n",
       " [1] parallel  stats4    grid      stats     graphics  grDevices utils    \n",
       " [8] datasets  methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] chromVAR_1.26.0                   motifmatchr_1.26.0               \n",
       " [3] chromVARmotifs_0.2.0              ggrepel_0.9.6                    \n",
       " [5] cowplot_1.1.3                     dplyr_1.1.4                      \n",
       " [7] BSgenome.Hsapiens.UCSC.hg38_1.4.5 BSgenome_1.72.0                  \n",
       " [9] rtracklayer_1.64.0                BiocIO_1.14.0                    \n",
       "[11] Biostrings_2.72.1                 XVector_0.44.0                   \n",
       "[13] patchwork_1.3.0                   rhdf5_2.48.0                     \n",
       "[15] SummarizedExperiment_1.34.0       Biobase_2.64.0                   \n",
       "[17] RcppArmadillo_14.2.3-1            Rcpp_1.0.14                      \n",
       "[19] Matrix_1.7-2                      GenomicRanges_1.56.2             \n",
       "[21] GenomeInfoDb_1.40.1               IRanges_2.38.1                   \n",
       "[23] S4Vectors_0.42.1                  BiocGenerics_0.50.0              \n",
       "[25] sparseMatrixStats_1.16.0          MatrixGenerics_1.16.0            \n",
       "[27] matrixStats_1.5.0                 data.table_1.17.0                \n",
       "[29] stringr_1.5.1                     plyr_1.8.9                       \n",
       "[31] magrittr_2.0.3                    gtable_0.3.6                     \n",
       "[33] gtools_3.9.5                      gridExtra_2.3                    \n",
       "[35] devtools_2.4.5                    usethis_3.1.0                    \n",
       "[37] ArchR_1.0.3                       ggplot2_3.5.0                    \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] jsonlite_1.9.0              farver_2.1.2               \n",
       "  [3] fs_1.6.5                    zlibbioc_1.50.0            \n",
       "  [5] vctrs_0.6.5                 memoise_2.0.1              \n",
       "  [7] Rsamtools_2.20.0            RCurl_1.98-1.16            \n",
       "  [9] base64enc_0.1-3             htmltools_0.5.8.1          \n",
       " [11] S4Arrays_1.4.1              curl_6.2.1                 \n",
       " [13] Rhdf5lib_1.26.0             CNEr_1.40.0                \n",
       " [15] SparseArray_1.4.8           htmlwidgets_1.6.4          \n",
       " [17] plotly_4.10.4               cachem_1.1.0               \n",
       " [19] uuid_1.2-1                  GenomicAlignments_1.40.0   \n",
       " [21] mime_0.12                   lifecycle_1.0.4            \n",
       " [23] pkgconfig_2.0.3             R6_2.6.1                   \n",
       " [25] fastmap_1.2.0               GenomeInfoDbData_1.2.13    \n",
       " [27] shiny_1.10.0                digest_0.6.37              \n",
       " [29] colorspace_2.1-1            TFMPvalue_0.0.9            \n",
       " [31] AnnotationDbi_1.66.0        pkgload_1.4.0              \n",
       " [33] RSQLite_2.3.9               seqLogo_1.70.0             \n",
       " [35] httr_1.4.7                  abind_1.4-8                \n",
       " [37] compiler_4.4.1              remotes_2.5.0              \n",
       " [39] bit64_4.6.0-1               withr_3.0.2                \n",
       " [41] BiocParallel_1.38.0         DBI_1.2.3                  \n",
       " [43] pkgbuild_1.4.6              R.utils_2.12.3             \n",
       " [45] poweRlaw_1.0.0              DelayedArray_0.30.1        \n",
       " [47] sessioninfo_1.2.3           rjson_0.2.23               \n",
       " [49] caTools_1.18.3              tools_4.4.1                \n",
       " [51] httpuv_1.6.15               R.oo_1.27.0                \n",
       " [53] glue_1.8.0                  restfulr_0.0.15            \n",
       " [55] rhdf5filters_1.16.0         promises_1.3.2             \n",
       " [57] pbdZMQ_0.3-13               reshape2_1.4.4             \n",
       " [59] TFBSTools_1.42.0            generics_0.1.3             \n",
       " [61] tzdb_0.4.0                  R.methodsS3_1.8.2          \n",
       " [63] nabor_0.5.0                 tidyr_1.3.1                \n",
       " [65] hms_1.1.3                   pillar_1.10.1              \n",
       " [67] IRdisplay_1.1               later_1.4.1                \n",
       " [69] lattice_0.22-6              bit_4.5.0.1                \n",
       " [71] annotate_1.82.0             tidyselect_1.2.1           \n",
       " [73] DirichletMultinomial_1.46.0 GO.db_3.20.0               \n",
       " [75] miniUI_0.1.1.1              DT_0.33                    \n",
       " [77] stringi_1.8.4               UCSC.utils_1.0.0           \n",
       " [79] lazyeval_0.2.2              yaml_2.3.10                \n",
       " [81] evaluate_1.0.3              codetools_0.2-20           \n",
       " [83] tibble_3.2.1                cli_3.6.4                  \n",
       " [85] IRkernel_1.3.2              xtable_1.8-4               \n",
       " [87] repr_1.1.7                  munsell_0.5.1              \n",
       " [89] png_0.1-8                   XML_3.99-0.18              \n",
       " [91] ellipsis_0.3.2              readr_2.1.5                \n",
       " [93] blob_1.2.4                  profvis_0.4.0              \n",
       " [95] urlchecker_1.0.1            bitops_1.0-9               \n",
       " [97] pwalign_1.0.0               viridisLite_0.4.2          \n",
       " [99] scales_1.3.0                purrr_1.0.4                \n",
       "[101] crayon_1.5.3                rlang_1.1.5                \n",
       "[103] KEGGREST_1.44.1            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "saveArchRProject(ArchRProj = proj, \n",
    "                 load = FALSE)\n",
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
